File: packages/domain/src/__tests__/incentive.test.ts
import { describe, expect, it } from 'vitest';
import type { Booking, ClassSession, IncentiveScheme } from '../index';
import {
  PAYOUT_ACTION_TARGET,
  PAYOUT_STATUSES,
  PAYOUT_TRANSITIONS,
  canTransition,
  computeCoachStatement,
  monthPeriod,
  periodMonthOf,
  resolveScheme,
} from '../index';

const NOW = '2026-08-20T10:00:00.000Z';

const defaultScheme: IncentiveScheme = {
  id: 'inc_default',
  coachId: null,
  sessionFeeIdr: 150_000,
  perAttendeeIdr: 10_000,
  fullClassBonusIdr: 100_000,
  fullClassThresholdPercent: 80,
  noShowPenaltyIdr: 0,
  rates: [],
  active: true,
  updatedAt: NOW,
};

const session = (id: string, over: Partial<ClassSession> = {}): ClassSession => ({
  id,
  classTypeId: 'cls_fund',
  branchId: 'brn_senopati',
  coachId: 'coa_1',
  startsAt: '2026-08-10T07:00:00.000Z',
  endsAt: '2026-08-10T08:00:00.000Z',
  capacity: 10,
  creditCost: 1,
  bookingOpensAt: '2026-08-03T07:00:00.000Z',
  bookingClosesAt: '2026-08-10T07:00:00.000Z',
  status: 'COMPLETED',
  area: null,
  ...over,
});

const booking = (sessionId: string, status: Booking['status'], n: number): Booking => ({
  id: `bok_${sessionId}_${status}_${n}`,
  memberId: `mem_${n}`,
  sessionId,
  status,
  waitlistPosition: null,
  source: 'MEMBER',
  createdAt: NOW,
  updatedAt: NOW,
  cancelledAt: null,
  checkedInAt: null,
  promotionOfferedAt: null,
});

const bookingsOf = (sessionId: string, counts: Partial<Record<Booking['status'], number>>) =>
  Object.entries(counts).flatMap(([status, count]) =>
    Array.from({ length: count }, (_, i) => booking(sessionId, status as Booking['status'], i)),
  );

const period = { start: '2026-08-01T00:00:00.000Z', end: '2026-09-01T00:00:00.000Z' };
const classTypesById = { cls_fund: { name: 'HYROX Fundamentals' } };

const compute = (
  sessions: ClassSession[],
  bookingsBySession: Record<string, Booking[]>,
  scheme: IncentiveScheme = defaultScheme,
) =>
  computeCoachStatement({
    coach: { id: 'coa_1' },
    sessions,
    bookingsBySession,
    classTypesById,
    scheme,
    period,
  });

describe('computeCoachStatement', () => {
  it('pays the session fee plus per-attendee for checked-in members only', () => {
    const s = session('ses_1');
    const res = compute([s], {
      ses_1: bookingsOf('ses_1', { CHECKED_IN: 3, COMPLETED: 2, NO_SHOW: 1, CANCELLED: 4 }),
    });
    expect(res.lines).toHaveLength(1);
    const line = res.lines[0]!;
    expect(line.booked).toBe(6); // cancelled never held a slot
    expect(line.attended).toBe(5);
    expect(line.noShows).toBe(1);
    expect(line.sessionFeeIdr).toBe(150_000);
    expect(line.attendeeIdr).toBe(50_000);
    expect(line.bonusIdr).toBe(0); // 5/10 = 50% < 80%
    expect(line.penaltyIdr).toBe(0);
    expect(line.totalIdr).toBe(200_000);
    expect(line.classTypeName).toBe('HYROX Fundamentals');
  });

  it('awards the full-class bonus exactly at the threshold', () => {
    const s = session('ses_1');
    const at = compute([s], { ses_1: bookingsOf('ses_1', { CHECKED_IN: 8 }) });
    expect(at.lines[0]!.bonusIdr).toBe(100_000);
    expect(at.lines[0]!.totalIdr).toBe(150_000 + 80_000 + 100_000);
    const below = compute([s], { ses_1: bookingsOf('ses_1', { CHECKED_IN: 7 }) });
    expect(below.lines[0]!.bonusIdr).toBe(0);
  });

  it('bonus counts attendance, not bookings', () => {
    const s = session('ses_1');
    const res = compute([s], { ses_1: bookingsOf('ses_1', { CONFIRMED: 9, CHECKED_IN: 1 }) });
    expect(res.lines[0]!.booked).toBe(10);
    expect(res.lines[0]!.bonusIdr).toBe(0);
  });

  it('applies the no-show penalty per no-show and never goes below zero', () => {
    const harsh = { ...defaultScheme, sessionFeeIdr: 50_000, noShowPenaltyIdr: 40_000 };
    const s = session('ses_1');
    const res = compute([s], { ses_1: bookingsOf('ses_1', { CHECKED_IN: 1, NO_SHOW: 3 }) }, harsh);
    const line = res.lines[0]!;
    expect(line.penaltyIdr).toBe(120_000);
    expect(line.totalIdr).toBe(0); // 50k + 10k − 120k → clamped
    expect(res.totals.totalIdr).toBe(0);
    expect(res.totals.penaltyIdr).toBe(120_000);
  });

  it('only counts COMPLETED sessions coached by this coach inside the period', () => {
    const sessions = [
      session('in'),
      session('other_coach', { coachId: 'coa_2' }),
      session('cancelled', { status: 'CANCELLED' }),
      session('upcoming', { status: 'PUBLISHED', startsAt: '2026-08-25T07:00:00.000Z' }),
      session('last_month', { startsAt: '2026-07-31T23:00:00.000Z' }),
      session('boundary_end', { startsAt: '2026-09-01T00:00:00.000Z' }),
      session('boundary_start', { startsAt: '2026-08-01T00:00:00.000Z' }),
    ];
    const res = compute(sessions, {});
    expect(res.lines.map((l) => l.sessionId)).toEqual(['boundary_start', 'in']);
    expect(res.totals.sessions).toBe(2);
    expect(res.totals.totalIdr).toBe(300_000); // fee only, no attendees
  });

  it('sums totals across lines', () => {
    const res = compute([session('a'), session('b', { startsAt: '2026-08-11T07:00:00.000Z' })], {
      a: bookingsOf('a', { CHECKED_IN: 8, NO_SHOW: 1 }),
      b: bookingsOf('b', { CHECKED_IN: 2 }),
    });
    expect(res.totals).toEqual({
      sessions: 2,
      attended: 10,
      noShows: 1,
      sessionFeeIdr: 300_000,
      attendeeIdr: 100_000,
      bonusIdr: 100_000,
      penaltyIdr: 0,
      totalIdr: 500_000,
    });
    expect(res.coachId).toBe('coa_1');
  });

  it('a zero-capacity session never earns the bonus', () => {
    const res = compute([session('z', { capacity: 0 })], { z: bookingsOf('z', { CHECKED_IN: 1 }) });
    expect(res.lines[0]!.bonusIdr).toBe(0);
  });
});

describe('resolveScheme', () => {
  const override: IncentiveScheme = {
    ...defaultScheme,
    id: 'inc_c1',
    coachId: 'coa_1',
    sessionFeeIdr: 200_000,
  };
  it('coach override wins when active', () => {
    expect(resolveScheme(defaultScheme, override).sessionFeeIdr).toBe(200_000);
  });
  it('falls back to the default without an override or when it is inactive', () => {
    expect(resolveScheme(defaultScheme, null)).toBe(defaultScheme);
    expect(resolveScheme(defaultScheme, { ...override, active: false })).toBe(defaultScheme);
  });
});

describe('payout state machine', () => {
  const legal: ReadonlyArray<readonly [string, string]> = [
    ['DRAFT', 'APPROVED'],
    ['DRAFT', 'VOID'],
    ['APPROVED', 'PAID'],
    ['APPROVED', 'VOID'],
  ];
  it('allows exactly the declared edges', () => {
    for (const from of PAYOUT_STATUSES) {
      for (const to of PAYOUT_STATUSES) {
        const expected = legal.some(([f, t]) => f === from && t === to);
        expect(canTransition(PAYOUT_TRANSITIONS, from, to), `${from} → ${to}`).toBe(expected);
      }
    }
  });
  it('PAID and VOID are terminal', () => {
    expect(PAYOUT_TRANSITIONS.PAID).toEqual([]);
    expect(PAYOUT_TRANSITIONS.VOID).toEqual([]);
  });
  it('every admin action targets a declared status', () => {
    for (const target of Object.values(PAYOUT_ACTION_TARGET))
      expect(PAYOUT_STATUSES).toContain(target);
    expect(PAYOUT_ACTION_TARGET).toEqual({ approve: 'APPROVED', pay: 'PAID', void: 'VOID' });
  });
});

describe('month periods', () => {
  it('produces an exclusive calendar-month window and round-trips the label', () => {
    const p = monthPeriod('2026-08');
    expect(new Date(p.start).getDate()).toBe(1);
    expect(new Date(p.end).getDate()).toBe(1);
    expect(new Date(p.end).getMonth()).toBe((new Date(p.start).getMonth() + 1) % 12);
    expect(periodMonthOf(p.start)).toBe('2026-08');
    expect(periodMonthOf(new Date(new Date(p.end).getTime() - 1).toISOString())).toBe('2026-08');
  });
  it('rolls over December', () => {
    const p = monthPeriod('2026-12');
    expect(new Date(p.end).getFullYear()).toBe(2027);
    expect(new Date(p.end).getMonth()).toBe(0);
  });
  it('rejects malformed labels', () => {
    expect(() => monthPeriod('2026-8')).toThrow();
  });
});

File: internal/domain/incentive_test.go
package domain

import (
	"testing"
	"time"
)

func testScheme() IncentiveScheme {
	return IncentiveScheme{
		ID:                        "sch_default",
		SessionFeeIDR:             150_000,
		PerAttendeeIDR:            15_000,
		FullClassBonusIDR:         50_000,
		FullClassThresholdPercent: 80,
		NoShowPenaltyIDR:          5_000,
		Active:                    true,
	}
}

func completedSession(id string, startsAt time.Time, capacity int) ClassSession {
	return ClassSession{
		ID: id, ClassTypeID: "clt_1", BranchID: "brn_1", CoachID: "coa_1",
		StartsAt: startsAt, EndsAt: startsAt.Add(time.Hour),
		Capacity: capacity, CreditCost: 1, Status: SessionCompleted,
	}
}

func bookings(statuses ...BookingStatus) []Booking {
	out := make([]Booking, 0, len(statuses))
	for i, s := range statuses {
		out = append(out, Booking{ID: string(rune('a' + i)), Status: s})
	}
	return out
}

func TestComputeCoachStatementPaysFeeAttendeesAndBonus(t *testing.T) {
	period := StatementPeriod{
		Start: time.Date(2026, 6, 1, 0, 0, 0, 0, time.UTC),
		End:   time.Date(2026, 7, 1, 0, 0, 0, 0, time.UTC),
	}
	session := completedSession("ses_1", time.Date(2026, 6, 10, 7, 0, 0, 0, time.UTC), 10)

	statement := ComputeCoachStatement(StatementInput{
		CoachID:  "coa_1",
		Sessions: []ClassSession{session},
		BookingsBySession: map[string][]Booking{
			// 8 of 10 attended: exactly the 80 percent bonus threshold.
			"ses_1": bookings(
				BookingCheckedIn, BookingCheckedIn, BookingCompleted, BookingCompleted,
				BookingCheckedIn, BookingCheckedIn, BookingCompleted, BookingCheckedIn,
				BookingNoShow, BookingCancelled,
			),
		},
		ClassTypeNames: map[string]string{"clt_1": "HYROX Fundamentals"},
		Scheme:         testScheme(),
		Period:         period,
	})

	if len(statement.Lines) != 1 {
		t.Fatalf("lines = %d, want 1", len(statement.Lines))
	}
	line := statement.Lines[0]
	if line.Attended != 8 {
		t.Fatalf("attended = %d, want 8", line.Attended)
	}
	if line.Booked != 9 {
		t.Fatalf("booked = %d, want 9 (attended plus the no-show)", line.Booked)
	}
	if line.NoShows != 1 {
		t.Fatalf("no-shows = %d, want 1", line.NoShows)
	}
	if line.BonusIDR != 50_000 {
		t.Fatalf("bonus = %d, want 50000 at exactly the threshold", line.BonusIDR)
	}
	// 150000 fee + 8*15000 attendees + 50000 bonus - 1*5000 penalty
	if line.TotalIDR != 315_000 {
		t.Fatalf("total = %d, want 315000", line.TotalIDR)
	}
	if line.ClassTypeName != "HYROX Fundamentals" {
		t.Fatalf("class name = %s", line.ClassTypeName)
	}
	if statement.Totals.TotalIDR != 315_000 || statement.Totals.Sessions != 1 {
		t.Fatalf("totals = %+v", statement.Totals)
	}
}

func TestComputeCoachStatementWithholdsBonusBelowThreshold(t *testing.T) {
	period := StatementPeriod{
		Start: time.Date(2026, 6, 1, 0, 0, 0, 0, time.UTC),
		End:   time.Date(2026, 7, 1, 0, 0, 0, 0, time.UTC),
	}
	statement := ComputeCoachStatement(StatementInput{
		CoachID:  "coa_1",
		Sessions: []ClassSession{completedSession("ses_1", time.Date(2026, 6, 10, 7, 0, 0, 0, time.UTC), 10)},
		BookingsBySession: map[string][]Booking{
			"ses_1": bookings(BookingCheckedIn, BookingCheckedIn, BookingCheckedIn),
		},
		Scheme: testScheme(),
		Period: period,
	})

	if statement.Lines[0].BonusIDR != 0 {
		t.Fatalf("bonus = %d, want 0 at 30 percent attendance", statement.Lines[0].BonusIDR)
	}
}

func TestComputeCoachStatementNeverPaysANegativeLine(t *testing.T) {
	period := StatementPeriod{
		Start: time.Date(2026, 6, 1, 0, 0, 0, 0, time.UTC),
		End:   time.Date(2026, 7, 1, 0, 0, 0, 0, time.UTC),
	}
	scheme := testScheme()
	scheme.SessionFeeIDR = 10_000
	scheme.NoShowPenaltyIDR = 50_000

	statement := ComputeCoachStatement(StatementInput{
		CoachID:  "coa_1",
		Sessions: []ClassSession{completedSession("ses_1", time.Date(2026, 6, 10, 7, 0, 0, 0, time.UTC), 10)},
		BookingsBySession: map[string][]Booking{
			"ses_1": bookings(BookingNoShow, BookingNoShow, BookingNoShow),
		},
		Scheme: scheme,
		Period: period,
	})

	if statement.Lines[0].TotalIDR != 0 {
		t.Fatalf("total = %d, want 0 (clamped, not negative)", statement.Lines[0].TotalIDR)
	}
}

func TestComputeCoachStatementCountsOnlyDeliveredSessionsInPeriodForThisCoach(t *testing.T) {
	period := StatementPeriod{
		Start: time.Date(2026, 6, 1, 0, 0, 0, 0, time.UTC),
		End:   time.Date(2026, 7, 1, 0, 0, 0, 0, time.UTC),
	}
	inside := completedSession("ses_in", time.Date(2026, 6, 15, 7, 0, 0, 0, time.UTC), 10)
	beforePeriod := completedSession("ses_before", time.Date(2026, 5, 31, 23, 0, 0, 0, time.UTC), 10)
	// The period end is exclusive, so a session at the boundary belongs to July.
	atEnd := completedSession("ses_end", time.Date(2026, 7, 1, 0, 0, 0, 0, time.UTC), 10)
	notCompleted := completedSession("ses_open", time.Date(2026, 6, 20, 7, 0, 0, 0, time.UTC), 10)
	notCompleted.Status = SessionPublished
	otherCoach := completedSession("ses_other", time.Date(2026, 6, 21, 7, 0, 0, 0, time.UTC), 10)
	otherCoach.CoachID = "coa_2"

	statement := ComputeCoachStatement(StatementInput{
		CoachID:  "coa_1",
		Sessions: []ClassSession{inside, beforePeriod, atEnd, notCompleted, otherCoach},
		Scheme:   testScheme(),
		Period:   period,
	})

	if len(statement.Lines) != 1 || statement.Lines[0].SessionID != "ses_in" {
		t.Fatalf("lines = %+v, want only ses_in", statement.Lines)
	}
}

func TestResolveSchemeFallsBackToTheOrganizationDefault(t *testing.T) {
	def := testScheme()
	coachID := "coa_1"
	override := testScheme()
	override.ID = "sch_coach"
	override.CoachID = &coachID
	override.SessionFeeIDR = 200_000

	if got := ResolveScheme(def, &override); got.ID != "sch_coach" {
		t.Fatalf("active override should win, got %s", got.ID)
	}

	override.Active = false
	if got := ResolveScheme(def, &override); got.ID != "sch_default" {
		t.Fatalf("inactive override should fall back to the default, got %s", got.ID)
	}
	if got := ResolveScheme(def, nil); got.ID != "sch_default" {
		t.Fatalf("missing override should fall back, got %s", got.ID)
	}
}

func TestMonthPeriodIsHalfOpen(t *testing.T) {
	period, err := MonthPeriod("2026-06", time.UTC)
	if err != nil {
		t.Fatalf("MonthPeriod returned %v", err)
	}
	if !period.Start.Equal(time.Date(2026, 6, 1, 0, 0, 0, 0, time.UTC)) {
		t.Fatalf("start = %s", period.Start)
	}
	if !period.End.Equal(time.Date(2026, 7, 1, 0, 0, 0, 0, time.UTC)) {
		t.Fatalf("end = %s, want the first of the next month", period.End)
	}
	if _, err := MonthPeriod("June 2026", time.UTC); err == nil {
		t.Fatal("a malformed period must be rejected")
	}
	if got := PeriodMonthOf(time.Date(2026, 6, 30, 23, 0, 0, 0, time.UTC), time.UTC); got != "2026-06" {
		t.Fatalf("period month = %s, want 2026-06", got)
	}
}

func TestPayoutTransitionsFollowTheApprovalChain(t *testing.T) {
	if _, err := Transition(PayoutTransitions, PayoutDraft, PayoutPaid); err == nil {
		t.Fatal("a draft payout must be approved before it can be paid")
	}
	if _, err := Transition(PayoutTransitions, PayoutApproved, PayoutPaid); err != nil {
		t.Fatalf("APPROVED -> PAID should be legal: %v", err)
	}
	if _, err := Transition(PayoutTransitions, PayoutPaid, PayoutVoid); err == nil {
		t.Fatal("a paid payout is final and cannot be voided")
	}
	if PayoutActionTarget[ActionApprove] != PayoutApproved {
		t.Fatal("approve must target APPROVED")
	}
}
